### PR TITLE
Optimize io (de)serialization routines

### PR DESCRIPTION
**Pull request by fyrchik**
Hiding `BinWriter` behind an interface allows us to use specialized implementations for buffer writer, counter writer etc.

`BinReader` can be impoved similarly, will do it later.

### CODE REVIEW

File: pkg/io/binaryBufWriter.go [hunk 1]
```diff
@@ -18,36 +18,141 @@ type BufferWriter interface {
...
+}
+
+func (bw *BufBinWriter) grow(n int) int {
+	ln := len(bw.buf)
+	if ln+n < cap(bw.buf) {
+		bw.buf = bw.buf[:ln+n]
+		return ln
+	}
+
+	buf := make([]byte, cap(bw.buf)*2+n)
```
Comment (AnnaShaleva) [inline pkg/io/binaryBufWriter.go:154]: Isn't it overhead a bit? Imagine that cap=100500, ln=100499 and n=1. Then new slice with a capacity=(cap*2+1) will be allocated.

Comment (fyrchik) [reply pkg/io/binaryBufWriter.go:154]: That's how it works everywhere (multiply by 2). This specific formula is taken from `bytes.Buffer`.

**Comment by fyrchik** [comment]
With `WriteArray` removal there are bigger relative gains:
```
name           old time/op    new time/op    delta
EncodeNew-8       614ns ± 6%     510ns ± 1%  -16.98%  (p=0.000 n=10+8)
EncodeReset-8     183ns ± 4%      87ns ± 3%  -52.52%  (p=0.000 n=10+10)
```

Another thing is new reader:
- calculate tx, block and extensible payload hash without re-serizalizing (see last commit, the interface can be improved)
- `ReadBytes` is a simple copy now (instead of intermediate slice allocation)

I think we can try to implement `Grow` and `VarSize` optimizations on top of existing `io` structures and wait with this massive refactoring until something else comes to mind.

**Comment by roman-khimov** [comment]
I think this one is severely outdated by now. Some recent optimizations also made it less relevant, so I'm closing it.

**Pull request closed without merging**